Sources/AWSLambdaRuntimeCore/LambdaHandler.swift

```diff
@@ -129,18 +129,42 @@ public protocol EventLoopLambdaHandler: ByteBufferLambdaHandler {
 public extension EventLoopLambdaHandler {
     /// Driver for `ByteBuffer` -> `In` decoding and `Out` -> `ByteBuffer` encoding
     func handle(context: Lambda.Context, event: ByteBuffer) -> EventLoopFuture<ByteBuffer?> {
-        switch self.decodeIn(buffer: event) {
+        let segment = context.tracer.beginSegment(name: "HandleEvent", baggage: context.baggage)
+        // TODO: record errors propagated in result types?
+        let decodedEvent = segment.subsegment(name: "DecodeIn") { _ in
+            self.decodeIn(buffer: event)
+        }
+        switch decodedEvent {
         case .failure(let error):
+            segment.addError(error)
+            segment.end()
             return context.eventLoop.makeFailedFuture(CodecError.requestDecoding(error))
         case .success(let `in`):
-            return self.handle(context: context, event: `in`).flatMapThrowing { out in
-                switch self.encodeOut(allocator: context.allocator, value: out) {
-                case .failure(let error):
-                    throw CodecError.responseEncoding(error)
-                case .success(let buffer):
-                    return buffer
+            // TODO: use NIO helpers?
+            let subsegment = segment.beginSubsegment(name: "HandleIn")
+            context.baggage = subsegment.baggage
+            return self.handle(context: context, event: `in`)
+                .always { result in
+                    if case .failure(let error) = result {
+                        subsegment.addError(error)
+                    }
+                    subsegment.end()
+                }
+                .flatMapThrowing { out in
+                    try context.tracer.segment(name: "EncodeOut", baggage: segment.baggage) { _ in
+                        switch self.encodeOut(allocator: context.allocator, value: out) {
+                        case .failure(let error):
+                            throw CodecError.responseEncoding(error)
+                        case .success(let buffer):
+                            return buffer
+                        }
+                    }
+                }.always { result in
+                    if case .failure(let error) = result {
+                        segment.addError(error)
+                    }
+                    segment.end()
```

Comment:
I wonder if this pattern can be made generic with some extensions to SwiftNIO futures

Reply:
Yup, let's kick off those "support" extension API discussions soon:  [Provide NIO Futures TracingSupport extensions package (not sure how to call the module yet) #125](https://github.com/slashmo/gsoc-swift-tracing/issues/125)

Reply:
updated the sugar in the XRaySDK API to make it cleaner and reusable